### PR TITLE
Storage initializer fix so that it downloads only specific file when provided uri is not a folder

### LINKED ISSUE #2473: S3 storage initializer doesn't work as expected for file without extension
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

I created the following files in my s3 bucket:
```
test/churn-pickle
test/churn-pickle-logs
test/churn-pickle-report
```
`churn-pickle`, `churn-pickle-logs` and `churn-pickle-report` are `file` without `extension`.

I use storage initializer to download `s3://<bucket>/test/churn-pickle` to `/mnt/models`. However, I got the following result:

```
❯ ls /mnt/models
total 24
-rw-r--r--  1 i543026  staff    12B 14 Oct 19:25 -logs
-rw-r--r--  1 i543026  staff    12B 14 Oct 19:25 -report
-rw-r--r--  1 i543026  staff    12B 14 Oct 19:25 churn-pickle
```

**What did you expect to happen:**

1. if the given path is a `file`, only the file is downloaded. In this case, `s3://<bucket>/test/churn-pickle` will be downloaded to `/mnt/models/churn-pickle`.
2. If the given path is a `folder`, all the files under the folder will be downloaded. 
3. If the given path is a `prefix` (not a folder or file), nothing should be downloaded.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

I assume that the storage initializer can only download from folders and files, not from `prefix`.

The issue was probably caused by:
1. objects are filter by the [prefix](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L167).
2. replace logic [here](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L167).

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version: 0.9
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`): v1.21.14
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by andyi2it**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2473 

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

**Feature/Issue validation/testing**:

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: python/kserve/kserve/storage/storage.py [hunk 1]
```diff
@@ -191,17 +192,24 @@ def _download_s3(uri, temp_dir: str):
...
-            target_key = (
-                obj.key.rsplit("/", 1)[-1]
-                if bucket_path == obj.key
-                else obj.key.replace(bucket_path, "", 1).lstrip("/")
-            )
+            if bucket_path == obj.key:
+                target_key = obj.key.rsplit("/", 1)[-1]
+                exact_obj_found = True
+            elif obj.key.rsplit("/", 1)[-1].startswith(bucket_path.rsplit("/", 1)[-1]):
+                target_key = obj.key.replace(bucket_path.rsplit("/", 1)[0], "", 1).lstrip("/")
```
Comment (yuzisun) [inline python/kserve/kserve/storage/storage.py]: @andyi2it Is this correct?  if `obj.key` is `churn-pickle-logs` and `bucket_path` is `churn-pickle`, then after replace it becomes `-logs`

Comment (yuzisun) [reply python/kserve/kserve/storage/storage.py]: Let's also add comment as the code is a bit convoluted here.

Comment (yuzisun) [reply python/kserve/kserve/storage/storage.py]: I think we need to define the expectation here when bucket is `s3://churn-pickle` and obj is `churn-pickle-logs/model.log`, do we still want to download all the files when there is a prefix matching or we do not match at all.

Comment (yuzisun) [reply python/kserve/kserve/storage/storage.py]: @lizzzcai Any thoughts?

Comment (lizzzcai) [reply python/kserve/kserve/storage/storage.py]: I feel like matching on a folder or exact file name rather than `prefix` will make things easier. But I am not sure if it will break any `existing user experience` here.


for bucket with following files:
```
test/churn-pickle
test/churn-pickle-logs
test/churn-pickle-report
```
1. When I download `s3://<bucket>/test/churn-pickle` to `/mnt/models`:
only one file is downloaded
```
/mnt/models/churn-pickle
```

2. when I download `s3://<bucket>/test` to `/mnt/models`:
All files under test will be downloaded.
```
/mnt/models/churn-pickle
/mnt/models/churn-pickle-logs
/mnt/models/churn-pickle-report
```

3. When I download `s3://<bucket>/test/churn` to `/mnt/models`:
nothing will be downloaded
```

```

not sure of any corner case here, and for cloud bucket storage, there is no easy way to know if a path is a file or folder.


@yuzisun @andyi2it any comments?

Comment (andyi2it) [reply python/kserve/kserve/storage/storage.py]: @yuzisun @lizzzcai 
Have refactored code to make it more readable.
With this PR, we won't download using prefixes only in case where an exact match is found.
So, for 3rd scenario you have, it should download
```
/mnt/models/churn-pickle
/mnt/models/churn-pickle-logs
/mnt/models/churn-pickle-report
```

This change should affect only the 1st scenario you have. When there is an exact match found, it will not use that as a prefix. This wouldn't or shouldn't break anything as without this PR, it would download the prefix matched files as below, which would not be usable in any case.

```
/mnt/models/-logs
/mnt/models/-report
```

Comment (lizzzcai) [reply python/kserve/kserve/storage/storage.py]: Hi @andyi2it,  for 1st scenario, there is a difference from the behaviour in ModelMesh [here](https://github.com/kserve/kserve/issues/2473#issuecomment-1715517612). In modelmesh, only `/mnt/models/churn-pickle` will be downloaded.
